lighthouse-core/computed/metrics/cumulative-layout-shift-all-frames.js

```diff
@@ -5,30 +5,70 @@
  */
 'use strict';
 
+/** @typedef {Omit<LH.TraceEvent, 'name'|'args'> & {name: 'LayoutShift', args: {data: {score: number, weighted_score_delta: number, had_recent_input: boolean}}}} LayoutShiftEvent */
+
 const makeComputedArtifact = require('../computed-artifact.js');
 const TraceOfTab = require('../trace-of-tab.js');
+const log = require('lighthouse-logger');
 
 class CumulativeLayoutShiftAllFrames {
+  /**
+   * @param {LH.TraceEvent} event
+   * @return {event is LayoutShiftEvent}
+   */
+  static isLayoutShiftEvent(event) {
+    return Boolean(
+      event.name === 'LayoutShift' &&
+      event.args &&
+      event.args.data &&
+      event.args.data.score !== undefined
+    );
+  }
+
   /**
    * @param {LH.Trace} trace
    * @param {LH.Audit.Context} context
    * @return {Promise<{value: number}>}
    */
   static async compute_(trace, context) {
     const traceOfTab = await TraceOfTab.request(trace, context);
-    const cumulativeShift = traceOfTab.frameTreeEvents
-      .filter(e =>
-        e.name === 'LayoutShift' &&
-        e.args &&
-        e.args.data &&
-        e.args.data.score &&
-        !e.args.data.had_recent_input
-      )
+    const layoutShiftEvents = traceOfTab.frameTreeEvents.filter(this.isLayoutShiftEvent);
+
+    // Chromium will set `had_recent_input` if there was recent user input, which
+    // skips shift events from contributing to CLS. This flag is also set when Lighthouse changes
+    // the emulation size. This consistently results in the first few shift event always being
+    // ignored for CLS. Since we don't expect any user input, we add the score of these
+    // shift events to CLS.
+    // See https://bugs.chromium.org/p/chromium/issues/detail?id=1094974.
+    for (const event of layoutShiftEvents) {
```

Comment:
Some more comments here would be useful. Is this accurate?

```
We set the first initial events that appear as `had_recent_input=true` to `false`, up until the first time `had_recent_input=false` appears. This ignores the incorrect inputs that are an artifact of Lighthouse changing the viewport size, and keeps any inputs that appear later in the page load (as a result of a puppeteer script controlling the page).
```

Reply:
does the logic here need to be more complex now in a multi-frame world? how does `had_recent_input` affect subframes when *their* viewports are resized from js? should those actually be counted? if so, the "had_recent_input" name is really starting to stretch 😆

Reply:
+1 to everything above, plus should we be doing this fixup in trace-of-tab? Or not altering the original trace events and doing it on a copy of the LayoutShift events in each of these audits...otherwise the state of early LayoutShift events depends on what other audits/computed artifacts have already run (e.g. an ads plugin CLS audit that either has to know about this issue or has to only run after the perf audits to get the numbers right)